### PR TITLE
gicv3: Disable reading GICR_TYPER for Aarch32

### PR DESCRIPTION
**Pull request by ayankuma**
GIC_TYPER is a 64 bit register. This is not supported on Aarch32 based platforms.

Signed-off-by: Ayan Kumar Halder <ayankuma@amd.com>

### CODE REVIEW

**Comment by stephanosio** [comment]
> Please ignore the commit "Merge branch 'main' into r52_fixes". I created this by mistake. I could not find a way to delete this.

Rebase and force-push

**Comment by ayankuma** [comment]
> Sorry, I am confused. Why does it not work for aarch32? I tested with `west build -p always -b fvp_baser_aemv8r_aarch32 zephyr/samples/synchronization/ -t run` (v8r aarch32 + gicv3) which works well for me. What am I missing? Could you please provide more info?

@povergoing  -
You are correct that it works on aarch32 FVP. However, looking at GIC v3 specification "Arm IHI 0069H ID020922", 12.11.37 GICR_TYPER, Redistributor Type Register, it is a 64 bit register with no equivalent 32 bit variant.

I am trying to run Zephyr for aarch32_v8r as a a DomU guest on Xen. The GIC redistributor region is emulated which means that when Zephyr will try to access GICR_TYPER, it will trap to Xen (with a data abort). Xen will read HSR = 0x9200000c (ARM DDI 0487G.b ID072021)).
This means :-
EC = 0b100100 (data abort from lower exception level)
IL = 0b1  (32-bit instruction trapped)
ISS = 0b1100 

Refer “ISS encoding for exception from a Data Abort ”, as ISV=0, thus ISS is invalid.
So Xen will not be able to emulate this instruction.

If we disable access to GICR_TYPER, there is no side-effect seen on the application. It seems GICR_TYPER has no use on aarch32_v8r.

**Comment by povergoing** [comment]
> Refer “ISS encoding for exception from a Data Abort ”, as ISV=0, thus ISS is invalid.
So Xen will not be able to emulate this instruction.
If we disable access to GICR_TYPER, there is no side-effect seen on the application. It seems GICR_TYPER has no use on aarch32_v8r.

@ayankuma I think I understand your issue. This might help you: https://community.arm.com/support-forums/f/soc-design-and-simulation-forum/52989/armv8-r-aem-fvp-trap-to-el2-missing-information-in-hsr-on-guest-s-strd. TBH, I prefer Xen could simulate this. Or if it turns out to be too hard, you could change `sys_read64` to 2 `sys_read32`s, just like linux. 

BTW, I remember there is a PR to do that adding something like `_gic_read64`, but I cannot find it. Kindly ping @carlocaione do you remember?

**Comment by povergoing** [comment]
> > Kindly ping @carlocaione do you remember?
> 
> I think that was when we were discussing [eb0428a](https://github.com/zephyrproject-rtos/zephyr/commit/eb0428a64a0b7ba32201fd53a53b2b03f154051e) but at the end we decided to have a specific named function for that instead of having a generic `gic_{read,write}64` helper.

Ah, it is, thanks. this way is good enough I think? @ayankuma would you like to change it to the 2 `sys_read32`s to read the GICR_TYPER when we are in CONFIG_ARM (AArch32) instead of this kind of hardcoding things?

**Comment by ayankuma** [comment]
> > > Kindly ping @carlocaione do you remember?
> > 
> > 
> > I think that was when we were discussing [eb0428a](https://github.com/zephyrproject-rtos/zephyr/commit/eb0428a64a0b7ba32201fd53a53b2b03f154051e) but at the end we decided to have a specific named function for that instead of having a generic `gic_{read,write}64` helper.
> 
> Ah, it is, thanks. this way is good enough I think? @ayankuma would you like to change it to the 2 `sys_read32`s to read the GICR_TYPER when we are in CONFIG_ARM (AArch32) instead of this kind of hardcoding things?

@povergoing Actually using 2 sys_read32() helped here as ISV = 1 (HSR=0x9383 000c), so the ISS is valid.
However, unlike Linux or running bare metal Zephyr, reading GICR_TYPER will trap into Xen (See [**here**](https://xenbits.xen.org/gitweb/?p=xen.git;a=blob;f=xen/arch/arm/vgic-v3.c;h=7fb99a9ff200c9d6735d2829bdf7b72dd2f7dae7;hb=9029bc265cdf2bd63376dde9fdd91db4ce9c0586#l191)). 

Xen expects to read and return GICR_TYPER as a uint64_t register.
For our solution, we need to modify Xen as below -

```
#ifdef AARCH32
case VREG32(GICR_TYPER_L32) :   
{
      read and return lower 32 bits of GICR_TYPER
      return 1;
}
case VREG32(GICR_TYPER_H32) :
{
      read and return the upper 32 bits of GICR_TYPER
      return 1;
}
#else /* For AARCH64 */
case VREG64(GICR_TYPER):
{
      do as existing;
}
#endif 
```

I will ask the Xen maintainers (also cc you) if such a solution seems acceptable or some better alternative exist.
Also, as there is no 32 bit variant of GICR_TYPER, thus I am not sure if enabling GICR_TYPER would make sense.

**Comment by jgrall** [comment]
> TBH, I prefer Xen could simulate this.

@povergoing Do you mean emulating ``strd``? If yes, this is not trivial and quite fragile. We would need to read the instruction from the guest memory and decode it. We also have to be careful because Xen read from the D-cache but the guest will execute the instruction from the I-Cache. This may be different value for various reason.

So I think it is better for Zephyr to use 32-bit access because this is going to be safer, faster and more portable (any hypervisor will have this issue).

**Comment by ayankuma** [comment]
> > TBH, I prefer Xen could simulate this.
> 
> @povergoing Do you mean emulating `strd`? If yes, this is not trivial and quite fragile. We would need to read the instruction from the guest memory and decode it. We also have to be careful because Xen read from the D-cache but the guest will execute the instruction from the I-Cache. This may be different value for various reason.
> 
> So I think it is better for Zephyr to use 32-bit access because this is going to be safer, faster and more portable (any hypervisor will have this issue).

@jgrall :- A minor correction. In this scenario, we are talking about emulating **ldrd**.

**Comment by povergoing** [comment]
> @povergoing Actually using 2 sys_read32() helped here as ISV = 1 (HSR=0x9383 000c), so the ISS is valid.
However, unlike Linux or running bare metal Zephyr, reading GICR_TYPER will trap into Xen (See [here](https://xenbits.xen.org/gitweb/?p=xen.git;a=blob;f=xen/arch/arm/vgic-v3.c;h=7fb99a9ff200c9d6735d2829bdf7b72dd2f7dae7;hb=9029bc265cdf2bd63376dde9fdd91db4ce9c0586#l191)).

Very happy it could work. Reading two times is no problem I think since it is only for booting. I think you should also support GICR_TYPER for AArch32 Linux on AArch32 Xen with GICv3 right? GICR_TYPER is a basic register for both A and R.

>@povergoing Do you mean emulating strd? If yes, this is not trivial and quite fragile. We would need to read the instruction from the guest memory and decode it. We also have to be careful because Xen read from the D-cache but the guest will execute the instruction from the I-Cache. This may be different value for various reason.

Really thanks for the explanation, clear to me.

>So I think it is better for Zephyr to use 32-bit access because this is going to be safer, faster and more portable (any hypervisor will have this issue).

Fair enough, using 32-bit access is totally good for me. 

@ayankuma but I still insist on reading GICR_TYPER even for v8R cause it is totally the same as v8A, I don't want to make it so special which could be a problem if v8r aarch32 wants to enable SMP in the future.

**Comment by ayankuma** [comment]
> > @povergoing Actually using 2 sys_read32() helped here as ISV = 1 (HSR=0x9383 000c), so the ISS is valid.
> > However, unlike Linux or running bare metal Zephyr, reading GICR_TYPER will trap into Xen (See [here](https://xenbits.xen.org/gitweb/?p=xen.git;a=blob;f=xen/arch/arm/vgic-v3.c;h=7fb99a9ff200c9d6735d2829bdf7b72dd2f7dae7;hb=9029bc265cdf2bd63376dde9fdd91db4ce9c0586#l191)).
> 
> Very happy it could work. Reading two times is no problem I think since it is only for booting. I think you should also support GICR_TYPER for AArch32 Linux on AArch32 Xen with GICv3 right? GICR_TYPER is a basic register for both A and R.
> 
> > @povergoing Do you mean emulating strd? If yes, this is not trivial and quite fragile. We would need to read the instruction from the guest memory and decode it. We also have to be careful because Xen read from the D-cache but the guest will execute the instruction from the I-Cache. This may be different value for various reason.
> 
> Really thanks for the explanation, clear to me.
> 
> > So I think it is better for Zephyr to use 32-bit access because this is going to be safer, faster and more portable (any hypervisor will have this issue).
> 
> Fair enough, using 32-bit access is totally good for me.
> 
> @ayankuma but I still insist on reading GICR_TYPER even for v8R cause it is totally the same as v8A, I don't want to make it so special which could be a problem if v8r aarch32 wants to enable SMP in the future.

@povergoing Many thanks for the review. I have pushed an updated patch with this change.

File: drivers/interrupt_controller/intc_gicv3.c
Comment (jgrall) [inline drivers/interrupt_controller/intc_gicv3.c:520]: Your proposed change reads a bit odd to me. You are meant to read the low and high 32-bit and then composed a 64-bit value. So I would expect the shift to be `32` rather than `GICR_TYPE_AFFINITY_VALUE_SHIFT`.

I understand the define will result to ``32`` as well. But that's less obvious. So I would suggest the following:

```
uint64_t val = sys_read32(rdist_addr + GICR_TYPER);
uint64_t gicr_aff;

val <<= 32;
val |= sys_read32(rdist_addr + GICR_TYPER + 4);

gicr_aff = GICR_TYPER_AFFINITY_VALUE_GET(val);
``` 

That said, I expect all the 64-bit access to GICv3 distributor/re-distributor to require two 32-bits accesses. Furthermore, IMHO, it would be better if the access stay 64-bit for Arm64. So shouldn't we provide a arm32 version of ``sys_read64()`` ?

Comment (ayankuma) [reply drivers/interrupt_controller/intc_gicv3.c:520]:
```
uint64_t val = sys_read32(rdist_addr + GICR_TYPER);
uint64_t gicr_aff;

val <<= 32;
val |= sys_read32(rdist_addr + GICR_TYPER + 4);

gicr_aff = GICR_TYPER_AFFINITY_VALUE_GET(val);
```
This seems incorrect to me as you are moving the value read from GICR_TYPER to the upper 32 bits. And the value read from (GICR_TYPER + 4) to the lower 32 bits.

> Furthermore, IMHO, it would be better if the access stay 64-bit for Arm64. So shouldn't we provide a arm32 version of sys_read64() ?
I was hoping that we could restrict splitting sys_read64() into 2 sys_read32() only for emulated MMIO regions (eg GICR, GICD). For other regions, it can stay as sys_read64(), isn't it ?

Comment (povergoing) [reply drivers/interrupt_controller/intc_gicv3.c:520]: Agreed with @jgrall idea, your proposed change seems to make `gicr_aff` only from the lower 32 bits, it would be wrong for AArch64.

Comment (jgrall) [reply drivers/interrupt_controller/intc_gicv3.c:520]: > This seems incorrect to me as you are moving the value read from GICR_TYPER to the upper 32 bits. And the value read from (GICR_TYPER + 4) to the lower 32 bits. It should be the other way around.

You are right.

> I was hoping that we could restrict splitting sys_read64() into 2 sys_read32() only for emulated MMIO regions (eg GICR, GICD). For other regions, it can stay as sys_read64(), isn't it ?

In theory any MMIO regions would have the same issue because the HW may not support 64-bit access on MMIO (see [1]). So I think they all should use two 32-bit accesses. 

[1] https://lore.kernel.org/xen-devel/6D2FD203-04D9-4CDD-B57D-E90F5BC5CC3A@arm.com/T/#mcc2b226e506c669918eee0266067f6d746ccc744

Comment (ayankuma) [reply drivers/interrupt_controller/intc_gicv3.c]: @jgrall @povergoing 
Please have a look at my latest patch.
I have modified sys_read64() into two sys_read32() for the aarch32 implementation.

Comment (povergoing) [reply drivers/interrupt_controller/intc_gicv3.c]: LGTM, but could it be more compact？ like
```
uint64_t val;

val = sys_read32(addr);
val |= (uint64_t)sys_read32(addr + 4) << 32;
```

**Comment by povergoing** [comment]
> In theory any MMIO regions would have the same issue because the HW may not support 64-bit access on MMIO (see [1]). So I think they all should use two 32-bit accesses.
> 
> [1] https://lore.kernel.org/xen-devel/6D2FD203-04D9-4CDD-B57D-E90F5BC5CC3A@arm.com/T/#mcc2b226e506c669918eee0266067f6d746ccc744

Kindly ping @microbuilder @carlocaione @stephanosio for comments, as this change has already been related to all `sys_read64` for AArch32.

**Comment by carlocaione** [comment]
> > In theory any MMIO regions would have the same issue because the HW may not support 64-bit access on MMIO (see [1]). So I think they all should use two 32-bit accesses.
> > [1] https://lore.kernel.org/xen-devel/6D2FD203-04D9-4CDD-B57D-E90F5BC5CC3A@arm.com/T/#mcc2b226e506c669918eee0266067f6d746ccc744
> 
> Kindly ping @microbuilder @carlocaione @stephanosio for comments, as this change has already been related to all `sys_read64` for AArch32.

We had already a similar discussion about this in #47171. As you can read in https://github.com/zephyrproject-rtos/zephyr/pull/47171#issuecomment-1176167605 all the `sys_{read,write}` operations are supposed to be atomic in Zephyr so we cannot really split a `sys_{read,write}64` in two 32-bit accesses without breaking atomicity.

We should resurrect that discussion (and maybe making the ball rolling again on that PR) but @quic-egmc was proposing to have a new set of MMIO helpers specifying the kind of access the arch is supporting.

**Comment by ayankuma** [comment]
> > > In theory any MMIO regions would have the same issue because the HW may not support 64-bit access on MMIO (see [1]). So I think they all should use two 32-bit accesses.
> > > [1] https://lore.kernel.org/xen-devel/6D2FD203-04D9-4CDD-B57D-E90F5BC5CC3A@arm.com/T/#mcc2b226e506c669918eee0266067f6d746ccc744
> > 
> > 
> > Kindly ping @microbuilder @carlocaione @stephanosio for comments, as this change has already been related to all `sys_read64` for AArch32.
> 
> We had already a similar discussion about this in #47171. As you can read in [#47171 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/47171#issuecomment-1176167605) all the `sys_{read,write}` operations are supposed to be atomic in Zephyr so we cannot really split a `sys_{read,write}64` in two 32-bit accesses without breaking atomicity.
> 
> We should resurrect that discussion (and maybe making the ball rolling again on that PR) but @quic-egmc was proposing to have a new set of MMIO helpers specifying the kind of access the arch is supporting.

@carlocaione This issue is related to 64 bit MMIO access on arm32 platforms **only**. AFAIU, #47171  is about 64 bit MMIO access on **all** architectures. Thus, I will prefer to keep them separate as the latter will require an agreement across all architecture. 
I am ok to rename `sys_read64` to `sys_read64_non_atomic` specifically for include/zephyr/arch/arm/aarch32/cortex_a_r/sys_io.h . Is it fine with you, @stephanosio, @microbuilder ?

**Comment by carlocaione** [comment]
> @carlocaione This issue is related to 64 bit MMIO access on arm32 platforms **only**. AFAIU, #47171 is about 64 bit MMIO access on **all** architectures. Thus, I will prefer to keep them separate as the latter will require an agreement across all architecture.

I do not agree. The two issues are very much related and if we can find a general solution it's very much preferred than solving this issue just for one architecture and having to fix this again when a general solution is found for all the architectures.

> I am ok to rename `sys_read64` to `sys_read64_non_atomic` specifically for include/zephyr/arch/arm/aarch32/cortex_a_r/sys_io.h . Is it fine with you, @stephanosio, @microbuilder ?

Not fine with me. This must be solved at a more general level.

**Comment by ayankuma** [comment]
@carlocaione @povergoing 
While reviewing #47171, I found an issue related to include/zephyr/arch/arm/aarch32/cortex_a_r/sys_io.h 

The functions sys_read/write() needs to be modified so that it uses inline assembly instructions to access MMIO regions. This is similar to how it is being done for arm64.

The reason for changing the arm32 code is :-
1. We want to avoid the compiler doing funny optimisations (for eg generating four ldrb instructions to read a 32 bit word). This may abort on certain MMIO registers which are to be accessed in 32 bit mode only.
2. Post indexing instructions for emulated MMIO access, are not supported when running on a hypervisor. Thus, we wish to prevent the compiler from generating any post indexing instructions.

See linux. commit id and the code is https://elixir.bootlin.com/linux/v6.1-rc1/source/arch/arm/include/asm/io.h
```
commit 195bbcac2e5c12f7fb99cdcc492c3000c5537f4a
Author: Will Deacon <will.deacon@arm.com>
Date:   Fri Aug 24 15:18:45 2012 +0100

    ARM: 7500/1: io: avoid writeback addressing modes for __raw_ accessors

    Data aborts taken to hyp mode do not provide a valid instruction
    syndrome field in the HSR if the faulting instruction is a memory
    access using a writeback addressing mode.
```

I can submit a change if it makes sense with you guys. 
I did mention the above issue in a comment on #47171 , but then realised it is better to fix it as a separate issue.

**Comment by carlocaione** [comment]
> I can submit a change if it makes sense with you guys.
> I did mention the above issue in a comment on #47171 , but then realised it is better to fix it as a separate issue.

Yeah, I agree. Open a new PR for this please, thanks.

**Comment by ayankuma** [comment]
> > I can submit a change if it makes sense with you guys.
> > I did mention the above issue in a comment on #47171 , but then realised it is better to fix it as a separate issue.
> 
> Yeah, I agree. Open a new PR for this please, thanks.

@carlocaione @povergoing  Many thanks, I have created #51403 to fix this.

**Comment by ayankuma** [comment]
@carlocaione There has not been any update on 

> > @carlocaione This issue is related to 64 bit MMIO access on arm32 platforms **only**. AFAIU, #47171 is about 64 bit MMIO access on **all** architectures. Thus, I will prefer to keep them separate as the latter will require an agreement across all architecture.
> 
> I do not agree. The two issues are very much related and if we can find a general solution it's very much preferred than solving this issue just for one architecture and having to fix this again when a general solution is found for all the architectures.
> 
> > I am ok to rename `sys_read64` to `sys_read64_non_atomic` specifically for include/zephyr/arch/arm/aarch32/cortex_a_r/sys_io.h . Is it fine with you, @stephanosio, @microbuilder ?
> 
> Not fine with me. This must be solved at a more general level.

@carlocaione There has not been any update on https://github.com/zephyrproject-rtos/zephyr/pull/47171 since the last 9 days. Do you suggest to wait longer or can you propose any way by which this PR can be unblocked ?

**Comment by carlocaione** [comment]
> @carlocaione There has not been any update on https://github.com/zephyrproject-rtos/zephyr/pull/47171 since the last 9 days. Do you suggest to wait longer or can you propose any way by which this PR can be unblocked ?

this is unfortunate :( do you think we can fix this locally following what was proposed in https://github.com/zephyrproject-rtos/zephyr/pull/51163#discussion_r997753439 ?

**Comment by povergoing** [comment]
> > @carlocaione There has not been any update on #47171 since the last 9 days. Do you suggest to wait longer or can you propose any way by which this PR can be unblocked ?
> 
> this is unfortunate :( do you think we can fix this locally following what was proposed in [#51163 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/51163#discussion_r997753439) ?

Might fix it with a specific function, just following the https://github.com/zephyrproject-rtos/zephyr/commit/eb0428a64a0b7ba32201fd53a53b2b03f154051e? And refine it when the https://github.com/zephyrproject-rtos/zephyr/pull/47171 is done?

**Comment by carlocaione** [comment]
> Might fix it with a specific function, just following the [eb0428a](https://github.com/zephyrproject-rtos/zephyr/commit/eb0428a64a0b7ba32201fd53a53b2b03f154051e)? And refine it when the #47171 is done?

Yes, I don't want to keep you blocked on this.

File: drivers/interrupt_controller/intc_gicv3.c
Comment (carlocaione) [inline drivers/interrupt_controller/intc_gicv3.c:516]: nit: remove empty line.

Comment (ayankuma) [reply drivers/interrupt_controller/intc_gicv3.c]: I have fixed this now.

**Review by carlocaione — APPROVED** [review]

**Comment by quic-egmc** [comment]
This is not a good solution.  With the #if defined() approach are injecting turmoil into drivers that must read 64-bit registers to do things two different ways - let's solve this with helper functions that capture the intent concisely rather than propagating architecture conditionals.

**Comment by ayankuma** [comment]
> This is not a good solution. With the #if defined() approach are injecting turmoil into drivers that must read 64-bit registers to do things two different ways - let's solve this with helper functions that capture the intent concisely rather than propagating architecture conditionals.

@quic-egmc There hasn't been any update on #47171 for a while now. What do you propose to unblock this PR ?

**Comment by povergoing** [comment]
> > > This is not a good solution. With the #if defined() approach are injecting turmoil into drivers that must read 64-bit registers to do things two different ways - let's solve this with helper functions that capture the intent concisely rather than propagating architecture conditionals.
> > 
> > 
> > @quic-egmc There hasn't been any update on #47171 for a while now. What do you propose to unblock this PR ?
> 
> I can update the PR to define the sys_read64_nonatomic but I wasn't clear that we had consensus since only @ruuddw responded and not @carlocaione or @andyross

Hi, @quic-egmc, we all understand #47171 is the right and more generic way. Thus it reasonably needs more time, right? :) That's why, finally, Ayan temporarily fixed this bug, and try not break or influence the 47171. But still, we all think it should be fixed with the helper as long as  #47171 is fixed.

**Comment by ayankuma** [comment]
> > > > This is not a good solution. With the #if defined() approach are injecting turmoil into drivers that must read 64-bit registers to do things two different ways - let's solve this with helper functions that capture the intent concisely rather than propagating architecture conditionals.
> > > 
> > > 
> > > @quic-egmc There hasn't been any update on #47171 for a while now. What do you propose to unblock this PR ?
> > 
> > 
> > I can update the PR to define the sys_read64_nonatomic but I wasn't clear that we had consensus since only @ruuddw responded and not @carlocaione or @andyross
> 
> Hi, @quic-egmc, we all understand #47171 is the right and more generic way. Thus it reasonably needs more time, right? :) That's why, finally, Ayan temporarily fixed this bug, and try not break or influence the 47171. But still, we all think it should be fixed with the helper as long as #47171 is fixed.

@povergoing I agree with you completely. I will humbly request you to merge this fix if it looks sane you so that we can run Zephyr as a DomU guest. 
Whenever #47171 gets merged, I promise to refix this.

**Review by povergoing — APPROVED** [review]
I give my +1 here, but we still need @quic-egmc's feedback or comments :)

**Comment by carlescufi** [comment]
Merging since this is -A and not -M.

**Pull request merged**
